File: backend/app/models/asyncapi.py
"""
event7 - AsyncAPI Models
Modèles pour la génération, le stockage et l'import des specs AsyncAPI.

Placement: backend/app/models/asyncapi.py

Changelog:
- v2: Added Kafka binding params (topic_name, partitions, replication_factor)
      Added include_key_schema flag for key/value separation
- v3: Added AsyncAPI import models (preview, request, result)
"""

from datetime import datetime

from pydantic import BaseModel, Field


# ════════════════════════════════════════════════════════════════════
# GENERATE
# ════════════════════════════════════════════════════════════════════


class AsyncAPIGenerateRequest(BaseModel):
    """Paramètres pour la génération d'une spec AsyncAPI."""

    # --- Existing params ---
    title: str | None = None
    description: str | None = None
    server_url: str | None = None
    include_examples: bool = True

    # --- Kafka binding params ---
    topic_name: str | None = None
    partitions: int | None = None
    replication_factor: int | None = None
    include_key_schema: bool = True

    # --- Confluent-specific params ---
    include_confluent_bindings: bool = True


class AsyncAPISpec(BaseModel):
    """Spec AsyncAPI stockée."""

    subject: str
    spec_content: dict
    is_auto_generated: bool = True
    updated_at: datetime | None = None


class AsyncAPIYamlExport(BaseModel):
    """YAML export wrapper."""

    content: str


# ════════════════════════════════════════════════════════════════════
# IMPORT — Preview & Apply
# ════════════════════════════════════════════════════════════════════


class AsyncAPIImportRequest(BaseModel):
    """Payload for import preview and apply."""

    spec_content: dict  # parsed AsyncAPI spec (JSON/dict, not YAML string)
    register_schemas: bool = False  # If true, push unknown schemas to SR


class ImportedChannel(BaseModel):
    """A channel extracted from the AsyncAPI spec."""

    address: str
    name: str
    broker_type: str = "kafka"
    resource_kind: str = "topic"
    messaging_pattern: str = "topic_log"
    data_layer: str | None = None
    description: str | None = None
    broker_config: dict = Field(default_factory=dict)


class ImportedBinding(BaseModel):
    """A subject binding extracted from a channel's messages."""

    channel_address: str
    subject_name: str
    schema_role: str = "value"
    binding_strategy: str = "channel_bound"
    binding_origin: str = "manual"
    binding_selector: str | None = None
    found_in_registry: bool = False  # True if subject exists in SR


class ImportedEnrichment(BaseModel):
    """Enrichment data extracted from the spec's info/extensions."""

    subject: str
    description: str | None = None
    owner_team: str | None = None
    tags: list[str] = Field(default_factory=list)
    data_layer: str | None = None


class ImportedSchema(BaseModel):
    """A schema found in the spec that is NOT in the SR."""

    subject_name: str
    schema_content: dict
    format: str = "JSON"  # AsyncAPI uses JSON Schema natively


class AsyncAPIImportPreview(BaseModel):
    """Dry-run result — what WOULD be created, nothing persisted yet."""

    # Spec metadata
    spec_title: str | None = None
    spec_version: str | None = None
    asyncapi_version: str | None = None

    # Extracted entities
    channels: list[ImportedChannel] = Field(default_factory=list)
    bindings: list[ImportedBinding] = Field(default_factory=list)
    enrichments: list[ImportedEnrichment] = Field(default_factory=list)
    unknown_schemas: list[ImportedSchema] = Field(default_factory=list)

    # Counts
    total_channels: int = 0
    total_bindings: int = 0
    total_enrichments: int = 0
    schemas_found: int = 0      # already in SR
    schemas_missing: int = 0    # not in SR

    # Warnings (non-blocking)
    warnings: list[str] = Field(default_factory=list)


class ImportEntityResult(BaseModel):
    """Result for a single entity creation."""

    entity_type: str  # "channel", "binding", "enrichment", "schema", "spec"
    name: str
    status: str  # "created", "updated", "skipped", "failed"
    detail: str | None = None


class AsyncAPIImportResult(BaseModel):
    """Result after applying the import."""

    # Counts
    channels_created: int = 0
    bindings_created: int = 0
    enrichments_updated: int = 0
    schemas_registered: int = 0
    spec_stored: bool = False

    # Detail log
    results: list[ImportEntityResult] = Field(default_factory=list)

    # Warnings
    warnings: list[str] = Field(default_factory=list)

File: frontend/src/types/asyncapi.ts
// src/types/asyncapi.ts
// TypeScript types for AsyncAPI feature
// v2: Added import types (preview, result, entities)

// ════════════════════════════════════════════════════════════════════
// GENERATE / VIEWER (existing)
// ════════════════════════════════════════════════════════════════════

export interface AsyncAPISpec {
  subject: string;
  spec_content: Record<string, unknown>;
  is_auto_generated: boolean;
  updated_at: string | null;
}

export interface AsyncAPIGenerateRequest {
  title?: string;
  description?: string;
  server_url?: string;
  include_examples?: boolean;
  topic_name?: string;
  partitions?: number;
  replication_factor?: number;
  include_key_schema?: boolean;
  include_confluent_bindings?: boolean;
}

export interface AsyncAPIYamlExport {
  subject: string;
  format: "yaml";
  content: string;
}

// ════════════════════════════════════════════════════════════════════
// IMPORT — Request, Preview, Result
// ════════════════════════════════════════════════════════════════════

export interface AsyncAPIImportRequest {
  spec_content: Record<string, unknown>;
  register_schemas?: boolean;
}

export interface ImportedChannel {
  address: string;
  name: string;
  broker_type: string;
  resource_kind: string;
  messaging_pattern: string;
  data_layer: string | null;
  description: string | null;
  broker_config: Record<string, unknown>;
}

export interface ImportedBinding {
  channel_address: string;
  subject_name: string;
  schema_role: string;
  binding_strategy: string;
  binding_origin: string;
  binding_selector: string | null;
  found_in_registry: boolean;
}

export interface ImportedEnrichment {
  subject: string;
  description: string | null;
  owner_team: string | null;
  tags: string[];
  data_layer: string | null;
}

export interface ImportedSchema {
  subject_name: string;
  schema_content: Record<string, unknown>;
  format: string;
}

export interface AsyncAPIImportPreview {
  spec_title: string | null;
  spec_version: string | null;
  asyncapi_version: string | null;
  channels: ImportedChannel[];
  bindings: ImportedBinding[];
  enrichments: ImportedEnrichment[];
  unknown_schemas: ImportedSchema[];
  total_channels: number;
  total_bindings: number;
  total_enrichments: number;
  schemas_found: number;
  schemas_missing: number;
  warnings: string[];
}

export interface ImportEntityResult {
  entity_type: string;
  name: string;
  status: "created" | "updated" | "skipped" | "failed";
  detail: string | null;
}

export interface AsyncAPIImportResult {
  channels_created: number;
  bindings_created: number;
  enrichments_updated: number;
  schemas_registered: number;
  spec_stored: boolean;
  results: ImportEntityResult[];
  warnings: string[];
}

// ════════════════════════════════════════════════════════════════════
// OVERVIEW (Dual Mode)
// ════════════════════════════════════════════════════════════════════

 
export interface SubjectAsyncAPIStatus {
  subject: string;
  origin: "imported" | "generated" | null;
  status: "documented" | "ready" | "raw";
  sync_status: "in_sync" | "outdated" | "unknown" | null;
  asyncapi_version: string | null;
  spec_title: string | null;
  spec_updated_at: string | null;
  governance_score: number | null;
  governance_grade: string | null;
  spec_version: number | null;
  has_enrichment: boolean;
  has_description: boolean;
  has_channels: boolean;
  has_bindings: boolean;
  description: string | null;
  owner_team: string | null;
  data_layer: string | null;
}
 
export interface AsyncAPIOverviewKPIs {
  total_subjects: number;
  documented: number;
  ready: number;
  raw: number;
  imported: number;
  generated: number;
  coverage_pct: number;
}
 
export interface AsyncAPIOverviewResponse {
  kpis: AsyncAPIOverviewKPIs;
  subjects: SubjectAsyncAPIStatus[];
}
 